pydatastructs/miscellaneous_data_structures/queue.py

```diff
@@ -364,3 +367,174 @@ def peek(self):
     @property
     def is_empty(self):
         return self.items.is_empty
+
+class ArrayDeque(Queue, DynamicOneDimensionalArray):
+    """
+    Represents Deque datastracture implemented using Array
+
+    Parameters
+    ==========
+
+    dtype: type
+        A valid object type.
+    size: int
+        The number of elements in the array.
+    elements: list/tuple
+        The elements in the array, all should
+        be of same type.
+    init: a python type
+        The inital value with which the element has
+        to be initialized. By default none, used only
+        when the data is not given.
+    load_factor: float, by default 0.25
+        The number below which if the ratio, Num(T)/Size(T)
+        falls then the array is contracted such that at
+        most only half the positions are filled.
+
+    Raises
+    ======
+
+    ValueError
+        When the number of elements in the list do not
+        match with the size.
+        More than three parameters are passed as arguments.
+        Types of arguments is not as mentioned in the docstring.
+        The load factor is not of floating point type.
+
+    IndexError
+        When pop/popleft is used on an empty Deque
+
+    Note
+    ====
+
+    At least one parameter should be passed as an argument along
+    with the dtype.
+    Num(T) means the number of positions which are not None in the
+    array.
+    Size(T) means the maximum number of elements that the array can hold.
+
+    Examples
+    ========
+
+    >>> from pydatastructs import Queue
+    >>> q = Queue(implementation='deque', dtype=int, items=[1,2,3])
+    >>> q.pop()
+    3
+    >>> q.popleft()
+    1
+    >>> q.append(1)
+    >>> q.appendleft(3)
+    >>> [q[i] for i in range(len(q))]
+    [3, 2, 1]
+    """
+
+    __slots__ = ['_load_factor', '_num', '_last_pos_filled', '_size', '_rear', '_front', '_num']
+
+    def __new__(cls, dtype, *args, **kwargs):
+        obj = DynamicOneDimensionalArray.__new__(cls, dtype, *args, **kwargs)
+        obj._front = obj._last_pos_filled
+        obj._rear = 0
+        obj._num = obj._front + 1
+        return obj
+
+    @property
+    def rear(self):
+        return (len(self._data) + self._rear) % len(self._data)
+
+    @property
+    def front(self):
+        return (len(self._data) + self._front) % len(self._data)
+
+    @front.setter
+    def front(self, val):
+        self._front = val
+
+    @rear.setter
+    def rear(self, val):
+        self._rear = val
+
+    @property
+    def size(self):
+        return self._num
+
+    @property
+    def is_empty(self):
+        return self.size == 0
```

Comment:
May be something using, `self.front` and `self.rear`. What's there in the parent classes? What's the MRO?

Reply:
tried using front and rear but this worked better since it is circular front can be behind rear in 2 cases if the deque is empty or if it is full and since % operator is used there is no way to find which one is the case